docs/x509.rst

```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
 
         The extensions encoded in the revoked certificate.
 
+X.509 CSR (Certificate Signing Request) Builder Object
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. class:: CertificateSigningRequestBuilder
+
+    .. versionadded:: 1.0
+
+    .. doctest::
+
+        >>> from cryptography import x509
+        >>> from cryptography.hazmat.backends import default_backend
+        >>> from cryptography.hazmat.primitives import hashes
+        >>> from cryptography.hazmat.primitives.asymmetric import rsa
+        >>> private_key = rsa.generate_private_key(
+        ...     public_exponent=65537,
+        ...     key_size=2048,
+        ...     backend=default_backend()
+        ... )
+        >>> builder = x509.CertificateSigningRequestBuilder()
+        >>> builder = builder.set_subject_name(x509.Name([
+        ...     x509.NameAttribute(x509.OID_COMMON_NAME, 'cryptography.io'),
+        ... ]))
+        >>> buidlder = builder.add_extension(x509.Extension(
+        ...     x509.OID_BASIC_CONSTRAINTS,
+        ...     True,
+        ...     x509.BasicConstraints(False, None),
+        ... ))
+        >>> request = builder.sign(
+        ...     default_backend(), private_key, hashes.SHA1()
+        ... )
+        >>> isinstance(request, x509.CertificateSigningRequest)
+        True
+
+    .. method:: __init__()
+
+        Creates an empty certificate signing request.
+
+    .. method:: set_subject_name(name)
```

Comment:
I'm of the opinion that Builders shouldn't have any getters.

On Sat, Jun 6, 2015 at 6:36 PM, André Caron <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1927#discussion_r31872134>:
>
> > +        >>> buidlder = builder.add_extension(x509.Extension(
> > +        ...     x509.OID_BASIC_CONSTRAINTS,
> > +        ...     True,
> > +        ...     x509.BasicConstraints(False, None),
> > +        ... ))
> > +        >>> request = builder.sign(
> > +        ...     default_backend(), private_key, hashes.SHA1()
> > +        ... )
> > +        >>> isinstance(request, x509.CertificateSigningRequest)
> > +        True
> > +
> > +    .. method:: __init__()
> > +
> > +        Creates an empty certificate signing request.
> > +
> > +    .. method:: set_subject_name(name)
>
> I don't mind changing this at all, but I'm wondering what the interface
> for eventual getter methods/properties would have..? If we rename
> set_subject_name() to subject_name(), then how would you call the getter?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927/files#r31872134>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084